### ISSUE TITLE
invalid genesis file: missing 0x prefix for hex data

### ISSUE THREAD

**Issue by kevinaquick** [ethereum/go-ethereum#14573]
Geth version: 1.6.5
OS & Version: Linux Ubuntu 14.04.5

We have a private network and am adding a node and I am getting the following error:

"Fatal: invalid genesis file: missing 0x prefix for hex data"  when calling

geth init /data/ethereum/customgenesis.json

Where customgenesis.json is our genesis data file that is used on all our nodes - all the hex data in this file is prefixed with 0x! and has worked in the past (the node was updated to geth 1.6.5 today)

(this error is actually happening on 2 nodes we are configuring both of which are on 1.6.5)

Any help appreciated

Kevin

**Comment by branciard** [comment]
Hi, was facing the same issue with my existing genesis file. Format seems to have changed. see  https://github.com/ethereum/go-ethereum/wiki/Private-network .
With the new format of the genesis file, geth init works.

**Comment by obscuren** [comment]
If you can provide me with your `genesis.json` I can have a look.

**Comment by kevinaquick** [comment]
Hi Thanks

The Genesis data is as follows:

{
    "nonce": "0x0000000000000042",
    "mixhash": "0x0000000000000000000000000000000000000000000000000000000000000000",
    "difficulty": "0x400",
    "alloc": {},
    "coinbase": "0x3333333333333333333333333333333333333333",
    "timestamp": "0x0",
    "parentHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
    "extraData": "Custom KMi blockchain Genesis Block",
    "gasLimit": "0x8000000"
}

**Comment by kevinaquick** [comment]
We have noticted if you change the extraData parameter to an empty string (or 0x0 I believe) then the genesis block can be initialised successfully i.e. without generating the "missing 0x prefix for hex data" error, BUT of course the node won't connect to the other running peers on the private network as there is a mismatch in the genesis block.

**Comment by holiman** [comment]
> BUT of course the node won't connect to the other running peers on the private network as there is a mismatch in the genesis block. 

If you specify the `extraData` as the hex-version of `Custom KMi blockchain Genesis Block`: `0x437573746f6d204b4d6920626c6f636b636861696e2047656e6573697320426c6f636b0a`  - does that work better?

**Comment by kevinaquick** [comment]
Hi 

I am afraid this still doesn't work (I spotted your edit) - I can initialise the genesis block (i.e. don't get the missing 0x error) but on running geth it fails to sync because of a genesis block mismatch:

DEBUG[06-08|09:59:44] Ethereum handshake failed                id=2ab1b2759ef91321 conn=inbound    err="Genesis block mismatch - 6650a0ac6c5e8054 (!= bbbdf3de44003a6f)"

**Comment by aphilsmith** [comment]
I got similar errors, but after using a blank extraData, the hex error goes away. I now see:

phil@etherium$ geth --datadir "$ethereum_home/youtube1" init "$ethereum_home/genesis.json"
WARN [06-10|17:55:37] No etherbase set and no accounts found as default 
INFO [06-10|17:55:37] Allocated cache and file handles         database=/android/etherium/youtube1/geth/chaindata cache=16 handles=16
Fatal: Failed to write genesis block: genesis has no chain configuration

**Comment by karalabe** [comment]
@kevinaquick 

Prior to Geth 1.6.0 (I think) the genesis loader didn't validate the fields you specified, just silently ignored any errors. However it always wanted the extra-data to be a hex string. In your particular case, `Custom KMi blockchain Genesis Block` failed to parse as a hex string, and was simply discarded and an empty extra-data used.

Running your sample genesis json through Geth 1.5.9 gets me:

```
$ cat genesis.json
{
"nonce": "0x0000000000000042",
"mixhash": "0x0000000000000000000000000000000000000000000000000000000000000000",
"difficulty": "0x400",
"alloc": {},
"coinbase": "0x3333333333333333333333333333333333333333",
"timestamp": "0x0",
"parentHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
"extraData": "Custom KMi blockchain Genesis Block",
"gasLimit": "0x8000000"
}

$geth --datadir=/tmp/a init genesis.json
cmd/geth/chaincmd.go:132] successfully wrote genesis block and/or chain rule set: 6231b02ac967dff9b0c799e956094408959de862d720d08776302ffefba0300b
```

Since the extraData was wrong and ignored, setting it to empty from Geth 1.6.0 onward should get you the same hash (running Geth 1.6.5):

```
$ cat genesis.json
{
"nonce": "0x0000000000000042",
"mixhash": "0x0000000000000000000000000000000000000000000000000000000000000000",
"difficulty": "0x400",
"alloc": {},
"coinbase": "0x3333333333333333333333333333333333333333",
"timestamp": "0x0",
"parentHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
"extraData": "0x",
"gasLimit": "0x8000000",
"config": {}
}

$ geth --datadir=/tmp/b init genesis.json
INFO [06-13|13:55:55] Successfully wrote genesis state         database=chaindata             hash=6231b0…a0300b
```

Which indeed happens.

TL;DR If you had invalid hex specified as the `extraData` until now, it was silently discarded. Please replace with `0x` and it will work fine.

---

@aphilsmith A custom genesis needs to specify a `config` section which defines when the protocol updates happen (Homestead, Spurious, Metropolis, etc). A trivial configuration is `"config": {}`, where no forks are applied, but I'd suggest running new private networks with all forks (you can generate such a genesis file via `puppeth` without needing to know the field names).

**Comment by rohitjana** [comment]
Hi

I replaced the extraData with 0x but I am still facing the same error message.

**Comment by karalabe** [comment]
What is "the same"?

**Comment by rohitjana** [comment]
geth --datadir data --networkid 123 --nodiscover --maxpeers 0 init g1.json
WARN [06-13|14:05:11] No etherbase set and no accounts found as default
INFO [06-13|14:05:11] Allocated cache and file handles         database=C:\\Users\\sysadmin\\Downloads\\geth-tutorial-master\\geth-tutorial-master\\nugget1\\data\\geth\\chaindata cache=16 handles=16
Fatal: Failed to write genesis block: genesis has no chain configuration

This is the error message i am facing. It similar to the posted by aphilsmith

**Comment by karalabe** [comment]
> @aphilsmith A custom genesis needs to specify a config section which defines when the protocol updates happen (Homestead, Spurious, Metropolis, etc). A trivial configuration is "config": {}, where no forks are applied, but I'd suggest running new private networks with all forks (you can generate such a genesis file via puppeth without needing to know the field names).

**Comment by rohitjana** [comment]
I understand now!

Thank you!!

**Comment by kevinaquick** [comment]
Hi Péter - thanks for the explanation re. the extradata field and the problem I was having. I have tried with it set to "0x" as suggested and this does indeed work and my node has now connected and synced with the nodes on the network. Thanks once again!  Kevin

**Comment by edoga7** [comment]
invalid genesis file: json: cannot unmarshal hex string without 0x prefix into Go struct field Genesis.extraData of data hexutil.Bytes

**Comment by pratikchhapolika** [comment]
I am getting the same error while running the genesis.bat file in windows 7:

"Fatal: invalid genesis file: json: cannot unmarshal hex string without 0x prefix into Go struct field Genesis.extraData of type hexutil.Bytes"
Genesis.bat conatins: "geth --datadir data --networkid 123 --nodiscover --maxpeers 0 init g1.json"

init contains: "@echo off
call clearchain
geth --datadir data --networkid 123 --nodiscover --maxpeers 0 init %1
echo "done."
"

Please help

**Comment by Princeredge** [comment]
good day guys are you familiar with genesis hack?
